File: python/build_full_network.py
"""
Build a fully-enhanced network in one pass, from any solvable input.

The pipeline is not tied to a particular case: it loads any network pypowsybl
can import (XIIDM, MATPOWER .mat/.m - with real base voltages kept automatically
- PSS/E .raw, CGMES, ...), runs a single AC load flow, and applies every
completion in this repo - reactive limits, ratio tap changers, a realistic
generation mix, apparent-power ratings, active power control, synthetic
measurements, observability flags and load-flow-based current + apparent-power
limit sets. By default it then rebuilds the result as a node-breaker network
(busbar sections + switches, with element properties carried across) and adds
the OpenLoadFlow outer-loop modeling that only node-breaker feeder bays can
carry (remote / shared / secondary / transformer / shunt voltage control, SVC
standby automaton, HVDC angle-droop control) so it looks like a real one.

The output exercises most of the IIDM object types and extensions, useful as a
realistic reference to compare synthetic networks against (see
network_summary.py).

Requirements and graceful degradation:
- The input must solve an AC load flow (a flat start is tried first, then a
  DC-based start); if neither converges the build aborts.
- Every completion operates on whatever equipment exists and reports counts, so
  features a given network lacks simply report zero (no EHV generators -> no
  remote voltage control, no multi-unit voltage levels -> no shared control, ...).
- The completions run on the (converging) bus-breaker network and the
  node-breaker conversion is done last, because the rebuilt topology loses the
  source's solved voltage profile and a flat start does not reconverge on very
  large cases (e.g. ACTIVSg70k). The node-breaker modeling steps each re-check
  convergence (DC init) and roll back if the rebuilt model does not reconverge,
  so on such a case they report zero while the fixture is still written. The
  converter carries operational limits and element-keyed extensions across; the
  measurements and observability extensions it does not copy are re-attached here.
"""

from __future__ import annotations

import argparse
import sys

import pypowsybl.loadflow as lf
import pypowsybl.network as pn

import add_current_limits as acl
import add_equipment as eq
import add_remote_voltage_control as rvc
import bus_to_node_breaker as b2nb
import complete_network as cn


def _reattach_measurements(src: pn.Network, dst: pn.Network) -> None:
    """Copy the measurements extension across a rebuild (converter skips it)."""
    df = src.get_extensions("measurements")
    if df.empty:
        return
    # pypowsybl transposes value/standard_deviation on create, so pass them
    # swapped; injection rows carry no side, branch rows do (a mixed empty/side
    # column trips the ThreeSides enum), so create the two groups separately.
    base = df.assign(value=df["standard_deviation"], standard_deviation=df["value"])
    inj = base[base["side"] == ""]
    if not inj.empty:
        dst.create_extensions("measurements", inj.drop(columns=["side"]))
    branch = base[base["side"] != ""]
    if not branch.empty:
        dst.create_extensions("measurements", branch)


def _reattach_observability(src: pn.Network, dst: pn.Network) -> None:
    """Copy the observability extensions across a rebuild (converter skips them)."""
    for name in ("injectionObservability", "branchObservability"):
        df = src.get_extensions(name)
        if df.empty:
            continue
        # Drop pypowsybl's internal *_null nullability markers before recreating.
        dst.create_extensions(name, df[[c for c in df.columns if not c.endswith("_null")]])


def _reattach_discrete_measurements(src: pn.Network, dst: pn.Network) -> None:
    """Copy the discreteMeasurements extension across a rebuild (element_id-keyed)."""
    df = src.get_extensions("discreteMeasurements")
    if not df.empty:
        dst.create_extensions("discreteMeasurements", df)


def _reattach_identifiable_short_circuit(src: pn.Network, dst: pn.Network) -> None:
    """Copy identifiableShortCircuit across a rebuild.

    Its dataframe carries a read-only ``equipment_type`` column that the create
    call rejects, so the generic converter skips it; drop that column and
    recreate here (the voltage levels keep their ids in the rebuild).
    """
    df = src.get_extensions("identifiableShortCircuit")
    if not df.empty:
        dst.create_extensions("identifiableShortCircuit",
                              df[[c for c in df.columns if c != "equipment_type"]])


def _load_case(input_path: str) -> pn.Network:
    """Load a case, keeping real base voltages for MATPOWER inputs.

    The MATPOWER importer flattens bus base voltages to 1.0 kV by default;
    ``matpower.import.ignore-base-voltage=false`` keeps the real nominal kV, which
    everything voltage-dependent (voltage class, short-circuit levels, remote-VC
    eligibility) relies on. Applied automatically for ``.mat``/``.m`` inputs.
    """
    if input_path.lower().endswith((".mat", ".m")):
        return pn.load(input_path,
                       parameters={"matpower.import.ignore-base-voltage": "false"})
    return pn.load(input_path)


def build_full(input_path: str, output_path: str,
               node_breaker: bool = True) -> pn.Network:
    """Apply every enhancement to the network at ``input_path`` and save it."""
    net = _load_case(input_path)

    # Inject synthetic equipment the source cases lack (batteries, SVCs, HVDC
    # links) before the load flow; they are created electrically neutral, so the
    # base case is unchanged.
    print(f"equipment               : {eq.add_equipment(net)}")

    cn._run_ac_or_raise(net, None)  # one load flow shared by every step

    steps = [
        ("reactive limits", lambda: cn.add_reactive_limits(net, run_loadflow=False)),
        ("ratio tap changers", lambda: cn.add_ratio_tap_changers(net, run_loadflow=False)),
        ("phase control", lambda: cn.add_phase_control(net)),
        ("svc voltage control", lambda: cn.add_svc_voltage_control(net)),
        ("generation mix", lambda: cn.set_generation_mix(net)),
        ("rated_s", lambda: cn.add_rated_s(net, run_loadflow=False)),
        ("reactive capability curves", lambda: cn.add_reactive_capability_curves(net)),
        ("active power control", lambda: cn.add_active_power_control(net)),
        ("short circuit", lambda: cn.add_short_circuit(net)),
        ("properties", lambda: cn.add_properties(net)),
        ("load detail", lambda: cn.add_load_detail(net)),
        ("measurements", lambda: cn.add_measurements(net, run_loadflow=False)),
        ("discrete measurements", lambda: cn.add_discrete_measurements(net)),
        ("observability", lambda: cn.add_observability(net)),
        ("current + apparent limits",
         lambda: acl.add_current_limits(net, run_loadflow=False,
                                        limit_types=("CURRENT", "APPARENT_POWER"))),
    ]
    for name, step in steps:
        result = step()
        print(f"{name:24s}: {result}")

    if node_breaker:
        # Rebuild as node-breaker last, then carry over the element_id-keyed
        # extensions the converter does not copy.
        nb = b2nb.convert(net)
        _reattach_measurements(net, nb)
        _reattach_observability(net, nb)
        _reattach_discrete_measurements(net, nb)
        _reattach_identifiable_short_circuit(net, nb)
        net = nb
        print(f"node-breaker            : {len(net.get_busbar_sections())} busbar "
              f"section(s), {len(net.get_switches())} switch(es)")

        # Remote voltage control: deport some EHV generators behind a new GSU
        # transformer (node-breaker, feeder bays) regulating the HV bus remotely.
        net, rvc_stats = rvc.deport_generators(net)
        print(f"remote voltage control  : {rvc_stats}")

        # Transformer voltage control: keep a converging subset of the ratio tap
        # changers regulating (thousands at once do not converge).
        print(f"transformer voltage ctl : {cn.add_transformer_voltage_control(net)}")

        # Secondary voltage control: a couple of control zones (pilot bus + a few
        # generators) - run before shared control so the zone generators are still
        # locally regulating.
        print(f"secondary voltage ctl   : {cn.add_secondary_voltage_control(net)}")

        # Shared voltage control: several generators in a voltage level co-regulate
        # one common bus.
        print(f"shared voltage control  : {cn.add_shared_voltage_control(net)}")

        # Shunt voltage control: a few switchable shunts regulating voltage.
        print(f"shunt voltage control   : {cn.add_shunt_voltage_control(net)}")

        # SVC standby automaton: put a converging subset of SVCs into standby
        # (fixed b0 until voltage leaves the band around its solved value).
        print(f"svc standby automaton   : {cn.add_svc_standby_automaton(net)}")

        # HVDC AC emulation: angle-droop active power control on the HVDC links.
        print(f"hvdc ac emulation       : {cn.add_hvdc_ac_emulation(net)}")

    # Solve the finished network with a DC-based voltage init so the fixture is
    # saved in (and reported as) a converged state. A flat start does not
    # converge on the rebuilt node-breaker model at this size (many retained
    # breakers); DC init converges the ~13k case. The very large node-breaker
    # case (ACTIVSg70k) does not reconverge from any non-solved start, so this is
    # reported, not enforced - the fixture is still written.
    params = lf.Parameters(distributed_slack=True, use_reactive_limits=True,
                           voltage_init_mode=lf.VoltageInitMode.DC_VALUES)
    status = lf.run_ac(net, params)[0].status.name
    print(f"final load flow (DC init): {status}")
    if status != "CONVERGED":
        print("  WARNING: fixture saved without a converged final state")

    net.save(output_path, format="XIIDM")
    print(f"Wrote {output_path}")
    return net


def _main(argv=None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("-i", "--input", required=True,
                        help="input case; a .mat/.m MATPOWER case is imported with "
                             "real base voltages automatically (e.g. case13659pegase.mat)")
    parser.add_argument("-o", "--output", default="full_network.xiidm.gz",
                        help="output network; a .gz suffix writes it compressed "
                             "(default: %(default)s)")
    parser.add_argument("--bus-breaker", action="store_true",
                        help="keep the bus-breaker topology (skip node-breaker conversion)")
    args = parser.parse_args(argv)
    build_full(args.input, args.output, node_breaker=not args.bus_breaker)
    return 0


if __name__ == "__main__":
    sys.exit(_main())

File: java/src/main/java/com/example/transporter/BuildFullNetwork.java
package com.example.transporter;

import com.powsybl.computation.local.LocalComputationManager;
import com.powsybl.ieeecdf.converter.IeeeCdfNetworkFactory;
import com.powsybl.iidm.network.ImportConfig;
import com.powsybl.iidm.network.Network;
import com.powsybl.iidm.serde.NetworkSerDe;
import com.powsybl.loadflow.LoadFlow;
import com.powsybl.loadflow.LoadFlowParameters;
import com.powsybl.loadflow.LoadFlowResult;

import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;
import java.util.function.Supplier;

/**
 * Build a fully-enhanced network in one pass — the Java port of
 * {@code build_full_network.py}.
 *
 * <p>Injects synthetic equipment, runs a load flow, applies every completion in
 * this repo (reactive limits, ratio tap changers, generation mix, rated_s,
 * reactive capability curves, active power control, short circuit, properties,
 * load detail, measurements, discrete measurements, observability, current +
 * apparent-power limits) and — by default — rebuilds the result as a
 * node-breaker network and adds the OpenLoadFlow outer-loop modeling (remote /
 * shared / secondary / transformer / shunt voltage control, SVC standby, HVDC
 * angle droop).
 *
 * <p>The source is either an input file or, when the first argument names one, a
 * bundled IEEE case ({@code ieee14}, {@code ieee57}, {@code ieee118},
 * {@code ieee300}) — the same set the Python CLI exposes via {@code --builtin}.
 *
 * <p>Unlike the Python pipeline (which completes on the bus-breaker network and
 * converts last), the typed IIDM API lets us convert first and complete on the
 * node-breaker network, so the extensions never have to be copied across the
 * rebuild. The base case must solve a load flow (a flat start is tried first,
 * then a DC start); the final load flow is reported, not enforced, so a case
 * whose node-breaker form does not reconverge is still written.
 */
public final class BuildFullNetwork {

    /** Bundled IEEE cases usable instead of an input file (mirrors the Python CLI). */
    private static final Map<String, Supplier<Network>> BUILTINS = new LinkedHashMap<>();

    static {
        BUILTINS.put("ieee14", IeeeCdfNetworkFactory::create14);
        BUILTINS.put("ieee57", IeeeCdfNetworkFactory::create57);
        BUILTINS.put("ieee118", IeeeCdfNetworkFactory::create118);
        BUILTINS.put("ieee300", IeeeCdfNetworkFactory::create300);
    }

    private BuildFullNetwork() {
    }

    private static void log(String name, Object stats) {
        System.out.printf("%-26s: %s%n", name, stats);
    }

    /**
     * Load a case, keeping real base voltages for MATPOWER inputs. The MATPOWER
     * importer flattens bus base voltages to 1 kV by default;
     * {@code matpower.import.ignore-base-voltage=false} keeps the real nominal kV,
     * which everything voltage-dependent relies on. Applied for .mat/.m inputs.
     */
    static Network loadCase(Path input) {
        Properties params = new Properties();
        String name = input.getFileName().toString().toLowerCase(Locale.ROOT);
        if (name.endsWith(".mat") || name.endsWith(".m")) {
            params.put("matpower.import.ignore-base-voltage", "false");
        }
        return Network.read(input, LocalComputationManager.getDefault(), ImportConfig.load(), params);
    }

    /** Apply every enhancement to {@code net} in place and return it. */
    public static Network buildFull(Network net, boolean nodeBreaker) {
        // Synthetic equipment (electrically neutral) before the load flow.
        log("equipment", EquipmentInjector.addEquipment(net));
        NetworkCompleter.runAcOrThrow(net);

        if (nodeBreaker) {
            net = BusToNodeBreakerConverter.convert(net);
            NetworkCompleter.runAcOrThrow(net);
            log("node-breaker", net.getBusbarSectionCount() + " busbar section(s), "
                    + net.getSwitchCount() + " switch(es)");
        }

        // Completions (each grounded on the solved base case).
        log("reactive limits", NetworkCompleter.addReactiveLimits(net,
                NetworkCompleter.ReactiveConfig.defaults().withRunLoadFlow(false)));
        log("ratio tap changers", NetworkCompleter.addRatioTapChangers(net,
                NetworkCompleter.RatioTapConfig.defaults().withRunLoadFlow(false)));
        log("generation mix", NetworkCompleter.setGenerationMix(net,
                NetworkCompleter.DEFAULT_GENERATION_MIX, true));
        log("rated_s", NetworkCompleter.addRatedS(net,
                NetworkCompleter.DEFAULT_GENERATOR_POWER_FACTOR,
                NetworkCompleter.DEFAULT_TRANSFORMER_LOADING, true, false));
        log("reactive capability curves", NetworkCompleter.addReactiveCapabilityCurves(net,
                NetworkCompleter.DEFAULT_CURVE_POINTS, true));
        log("active power control", NetworkCompleter.addActivePowerControl(net,
                NetworkCompleter.DEFAULT_DROOP, true));
        log("short circuit", NetworkCompleter.addShortCircuit(net, true));
        log("properties", NetworkCompleter.addProperties(net,
                NetworkCompleter.DEFAULT_REGION_COUNT, NetworkCompleter.DEFAULT_COUNTRY, true));
        log("load detail", NetworkCompleter.addLoadDetail(net,
                NetworkCompleter.DEFAULT_LOAD_FIXED_FRACTION, true));
        log("measurements", NetworkCompleter.addMeasurements(net,
                NetworkCompleter.DEFAULT_MEASUREMENT_STD_DEV, NetworkCompleter.DEFAULT_STD_DEV_FLOOR,
                true, false));
        log("discrete measurements", NetworkCompleter.addDiscreteMeasurements(net, true));
        log("observability", NetworkCompleter.addObservability(net,
                NetworkCompleter.DEFAULT_OBSERVABILITY_STD_DEV, true));
        log("current + apparent limits", CurrentLimitsGenerator.apply(net,
                CurrentLimitsGenerator.Config.defaults().withRunLoadFlow(false).withLimitTypes(
                        java.util.List.of(com.powsybl.iidm.network.LimitType.CURRENT,
                                com.powsybl.iidm.network.LimitType.APPARENT_POWER))));

        if (nodeBreaker) {
            // OLF outer-loop modeling (each keeps the base case converged).
            RemoteVoltageControl.DeportResult rvc = RemoteVoltageControl.deportGenerators(
                    net, null, RemoteVoltageControl.DEFAULT_HV_THRESHOLD,
                    RemoteVoltageControl.DEFAULT_LV_NOMINAL_V, RemoteVoltageControl.DEFAULT_GSU_X_PU);
            net = rvc.network();
            log("remote voltage control", "deported " + rvc.deported() + "/" + rvc.eligible());
            log("transformer voltage ctl", VoltageControlCompleter.addTransformerVoltageControl(
                    net, VoltageControlCompleter.DEFAULT_TVC_COUNT,
                    VoltageControlCompleter.DEFAULT_TVC_DEADBAND));
            log("secondary voltage ctl", VoltageControlCompleter.addSecondaryVoltageControl(
                    net, VoltageControlCompleter.DEFAULT_SECONDARY_VC_ZONES,
                    VoltageControlCompleter.DEFAULT_SECONDARY_VC_UNITS));
            log("shared voltage control", VoltageControlCompleter.addSharedVoltageControl(
                    net, VoltageControlCompleter.DEFAULT_SHARED_VC_GROUPS));
            log("shunt voltage control", VoltageControlCompleter.addShuntVoltageControl(
                    net, VoltageControlCompleter.DEFAULT_SHUNT_VC_COUNT,
                    VoltageControlCompleter.DEFAULT_SHUNT_SECTIONS,
                    VoltageControlCompleter.DEFAULT_SHUNT_B_PER_SECTION, 1.0));
            log("phase control", VoltageControlCompleter.addPhaseControl(
                    net, null, VoltageControlCompleter.DEFAULT_PHASE_CURRENT_MARGIN));
            log("svc voltage control", VoltageControlCompleter.addSvcVoltageControl(
                    net, VoltageControlCompleter.DEFAULT_SVC_SLOPE));
            log("svc standby automaton", VoltageControlCompleter.addSvcStandbyAutomaton(net, null));
            log("hvdc ac emulation", VoltageControlCompleter.addHvdcAcEmulation(
                    net, VoltageControlCompleter.DEFAULT_HVDC_DROOP, null));

            // Java-only modeling (reachable in powsybl-core but not pypowsybl).
            log("generator remote reactive", JavaOnlyModeling
                    .addGeneratorRemoteReactivePowerControl(net, 1));
            log("transformer reactive ctl", JavaOnlyModeling
                    .addTransformerReactivePowerControl(net, 1));
            log("automation systems", JavaOnlyModeling.addAutomationSystems(net, 1));
        }

        // Solve with a DC-based init so the fixture is saved converged; reported,
        // not enforced (a very large node-breaker case may not reconverge).
        LoadFlowParameters params = new LoadFlowParameters()
                .setDistributedSlack(true).setUseReactiveLimits(true)
                .setVoltageInitMode(LoadFlowParameters.VoltageInitMode.DC_VALUES);
        LoadFlowResult result = LoadFlow.run(net, params);
        log("final load flow (DC init)", result.getStatus());
        if (!result.isFullyConverged()) {
            System.out.println("  WARNING: fixture saved without a converged final state");
        }
        return net;
    }

    /**
     * Resolve the source case: a bundled IEEE network when {@code source} names one
     * ({@code ieee14}, {@code ieee57}, {@code ieee118}, {@code ieee300}), else the
     * file at that path.
     */
    static Network resolveCase(String source) {
        Supplier<Network> builtin = BUILTINS.get(source.toLowerCase(Locale.ROOT));
        return builtin != null ? builtin.get() : loadCase(Path.of(source));
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("usage: BuildFullNetwork <input|" + String.join("|", BUILTINS.keySet())
                    + "> <output> [--bus-breaker]");
            System.exit(2);
        }
        boolean nodeBreaker = args.length < 3 || !"--bus-breaker".equals(args[2]);
        Network net = resolveCase(args[0]);
        net = buildFull(net, nodeBreaker);
        NetworkSerDe.write(net, Path.of(args[1]));
        System.out.println("Wrote " + args[1]);
    }
}
